bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/RuleEngine.java

```diff
@@ -0,0 +1,426 @@
+                    result.add(new RuleImpl(r));
+                }
+            } else {
+                result.add(new RuleImpl(r));
+            }
+        }
+        return result;
+    }
+
+    public Collection<Rule> getRules(String filter) {
+        Collection<Rule> result = new ArrayList<Rule>(10);
+        for (Iterator<RuleImpl> it = rules.values().iterator(); it.hasNext();) {
+            RuleImpl r = it.next();
+            result.add(new RuleImpl(r));
+        }
+        return result;
+    }
+
+    protected Collection<String> getScopeIds() {
+        Set<String> result = new HashSet<String>(10);
+        for (Iterator<RuleImpl> it = rules.values().iterator(); it.hasNext();) {
+            RuleImpl r = it.next();
+            String id = r.getScopeIdentifier();
+            if (id != null) {
+                result.add(id);
+            }
+        }
+        return result;
+    }
+
+    /**
+     *
+     * @see org.osgi.util.tracker.ServiceTrackerCustomizer#addingService(org.osgi.framework.ServiceReference)
+     */
+    @Override
+    public ModuleHandlerFactory addingService(ServiceReference/* <ModuleHandlerFactory> */reference) {
+        ModuleHandlerFactory msf = (ModuleHandlerFactory) bc.getService(reference);
+        Collection<String> moduleTypes = msf.getTypes();
+        for (Iterator<String> it = moduleTypes.iterator(); it.hasNext();) {
+            String moduleTypeName = it.next();
+            if (moduleHandlerFactories.get(moduleTypeName) != null) {
+                // log.error
+            } else {
+                moduleHandlerFactories.put(moduleTypeName, msf);
+            }
+
+        }
+        return msf;
+    }
+
+    /**
+     *
+     * @see org.osgi.util.tracker.ServiceTrackerCustomizer#modifiedService(org.osgi.framework.ServiceReference,
+     *      java.lang.Object)
+     */
+    @Override
+    public void modifiedService(ServiceReference/* <ModuleHandlerFactory> */reference, /* ModuleHandlerFactory */
+            Object service) {
+        // TODO Auto-generated method stub
+
+    }
+
+    /**
+     * @see org.osgi.util.tracker.ServiceTrackerCustomizer#removedService(org.osgi.framework.ServiceReference,
+     *      java.lang.Object)
+     */
+    @Override
+    public void removedService(ServiceReference/* <ModuleHandlerFactory> */reference, /* ModuleHandlerFactory */
+            Object service) {
+        Collection<String> moduleTypes = ((ModuleHandlerFactory) service).getTypes();
+        for (Iterator<String> it = moduleTypes.iterator(); it.hasNext();) {
+            String moduleTypeName = it.next();
+            moduleHandlerFactories.remove(moduleTypeName);
+        }
+    }
+
+    public void runRule(RuleEngineCallbackImpl reCallback) {
+        synchronized (reCallback) {
+            Rule r = reCallback.getRule();
+            Thread t = runningRules.get(r.getUID());
+
+            if (t == null) {
+                RuleThread rt = new RuleThread(reCallback); // TODO use ThreadPool
```

Comment:
> We don't use thread pool because it is part of prosyst apis. 

This should not prevent us from using a thread pool here. We can externalize it in a way that you can plug in your own version. E.g. for discovery services we have 
https://github.com/eclipse/smarthome/blob/master/bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/util/DiscoveryThreadPool.java. What is provided is a ScheduledExecutorService - we could imho do the same here.

> The trigger handled does not start the rule.

This is clear. What I meant was providing a scheduler that can (must!) be used within TriggerHandler implementations.  We have this e.g. for ThingHandlers as well (https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java#L48). This should actually go into the automation.api bundle, not into the automation.core.